node/coinstacks/solana/api/src/websocket.ts

```diff
@@ -0,0 +1,107 @@
+import { Logger } from '@shapeshiftoss/logger'
+import { WebsocketClient as BaseWebsocketClient, Args, Options, Subscription } from '@shapeshiftoss/websocket'
+import WebSocket from 'ws'
+import { Transaction, WebsocketResponse } from './types'
+
+const logger = new Logger({ namespace: ['unchained', 'solana', 'websocket'], level: process.env.LOG_LEVEL })
+
+interface WebsocketArgs extends Omit<Args, 'logger'> {
+  transactionHandler: TransactionHandler | Array<TransactionHandler>
+}
+
+type TransactionHandler = (data: Transaction) => Promise<void>
+
+export class WebsocketClient extends BaseWebsocketClient {
+  private handleTransaction: TransactionHandler | Array<TransactionHandler>
+
+  private addresses: Array<string> = []
+  private subscriptionsIds: Array<number> = []
+
+  constructor(_url: string, args: WebsocketArgs, opts?: Options) {
+    const url = args.apiKey ? `${_url}?api-key=${args.apiKey}` : _url
+    super(url, { logger }, opts)
+
+    this.handleTransaction = args.transactionHandler
+
+    this.initialize()
+
+    this.socket.onmessage = (msg) => this.onMessage(msg)
+    this.socket.onopen = () => this.onOpen()
+  }
+
+  protected onOpen(): void {
+    super._onOpen()
+
+    if (this.addresses.length > 0) {
+      const subscribeAddresses = this.getAddressesSubscription()
+      this.socket.send(JSON.stringify(subscribeAddresses))
+    }
+  }
+
+  protected async onMessage(message: WebSocket.MessageEvent): Promise<void> {
+    try {
+      const res: WebsocketResponse = JSON.parse(message.data.toString())
+
+      if (res.params?.subscription && !this.subscriptionsIds.includes(res.params.subscription)) {
```

Comment:
there is a subscription result that comes back first that we can narrow on `{"jsonrpc":"2.0","result":8688172533891622,"id":"newTx"}` for the id to reduce the need for running the `includes` on every message the comes accross.

Reply:
Fixed in [87e2e9c](https://github.com/shapeshift/unchained/pull/1053/commits/87e2e9c40e6546c7ce72cb7413ae8f1175b16529)